jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/BufferedContentSink.java

```diff
@@ -0,0 +1,297 @@
+    private boolean _lastWritten;
+    private volatile CountingCallback _countingCallback;
+
+    private enum AggregateState
+    {
+        IDLE,
+        WRITING,
+        ITERATING,
+        PENDING
+    }
+
+    public BufferedContentSink(Content.Sink delegate, ByteBufferPool bufferPool, boolean direct, int maxBufferSize, int maxAggregationSize)
+    {
+        if (maxBufferSize <= 0)
+            throw new IllegalArgumentException("maxBufferSize must be > 0, was: " + maxBufferSize);
+        if (maxAggregationSize <= 0)
+            throw new IllegalArgumentException("maxAggregationSize must be > 0, was: " + maxAggregationSize);
+        if (maxBufferSize < maxAggregationSize)
+            throw new IllegalArgumentException("maxBufferSize (" + maxBufferSize + ") must be >= maxAggregationSize (" + maxAggregationSize + ")");
+        _delegate = delegate;
+        _bufferPool = (bufferPool == null) ? new ByteBufferPool.NonPooling() : bufferPool;
+        _direct = direct;
+        _maxBufferSize = maxBufferSize;
+        _maxAggregationSize = maxAggregationSize;
+    }
+
+    @Override
+    public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("writing last={} {}", last, BufferUtil.toDetailString(byteBuffer));
+
+        if (_lastWritten)
+        {
+            callback.failed(new IOException("complete"));
+            return;
+        }
+        _lastWritten = last;
+        if (_firstWrite)
+        {
+            _firstWrite = false;
+            if (last)
+            {
+                // No need to buffer if this is both the first and the last write.
+                _delegate.write(true, byteBuffer, callback);
+                return;
+            }
+        }
+
+        ByteBuffer current = byteBuffer != null ? byteBuffer : BufferUtil.EMPTY_BUFFER;
+        if (current.remaining() > _maxAggregationSize)
+        {
+            // current buffer is greater than the max aggregation size
+            directWrite(last, current, callback);
+            return;
+        }
+
+        // current buffer can be aggregated
+        if (_aggregator == null)
+            _aggregator = new ByteBufferAggregator(_bufferPool, _direct, Math.min(START_BUFFER_SIZE, _maxBufferSize), _maxBufferSize);
+
+        /*
+         * The given callback might call write() again, which could stack overflow.
+         * To avoid this, we must keep a state to avoid re-entering write() when the
+         * callback is succeeded by the delegate's write() call and delay the given
+         * callback's success to when we're exiting the original write() call.
+         */
+        AggregateState state = _aggregateState.updateAndGet(s -> switch (s)
+        {
+            case IDLE -> AggregateState.WRITING;
+            case ITERATING -> AggregateState.ITERATING;
+            case WRITING, PENDING -> throw new WritePendingException();
+        });
```

Comment:
What? Isn't this `IteratingCallback` rewritten?

Reply:
@lorban @sbordet I don't think this is a job for iterating callback, it is similar but different.

Specifically there is a handoff from one thread to the other of a callback, but not an unlimited one.   
The second thread writing needs to know for sure that the first thread writing is going to complete its callback, else it becomes the writing thread.  If we turn that into a queue and iterate operation, then firstly we have a queue when we really don't need one, and secondly we now allow N concurrent writes, which we don't want to.

But yeah try to write it with a persistent iterating callback and see the difference.

Reply:
Done.